proposals/cleanup.md

```diff
@@ -0,0 +1,256 @@
+# Meta
+
+- Name: Clean-up
+- Start Date: 2022-07-30
+- Author(s): chipzoller
+- Supersedes: N/A
+
+# Table of Contents
+
+- [Meta](#meta)
+- [Table of Contents](#table-of-contents)
+- [Overview](#overview)
+- [Definitions](#definitions)
+- [Motivation](#motivation)
+- [Proposal](#proposal)
+- [Implementation](#implementation)
+- [Migration (OPTIONAL)](#migration-optional)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Prior Art](#prior-art)
+- [Unresolved Questions](#unresolved-questions)
+- [CRD Changes (OPTIONAL)](#crd-changes-optional)
+
+# Overview
+
+A clean-up ability of Kyverno allowing it to clean-up (delete) resources in a Kubernetes cluster based upon certain criteria.
+
+# Definitions
+
+- TTL: Time-to-live. The amount of time a resource may exist until it is cleaned up.
+
+# Motivation
+
+The initial idea for this feature came from issue [3483](https://github.com/kyverno/kyverno/issues/3483) authored by user luisdavim.
+
+As long as there has been Kubernetes, there has been the inevitability that some resources--most especially those created directly by human users--are left derelict at some point without a good way to identify what and where they are. As Kubernetes continues to expand with its myriad of possible custom resources, this problem grows over time. The ultimate problems this poses are few: 1) Kubernetes' key-value store, which has fairly limited storage capacity, may become overwhelmed by resources which should be deleted; and 2) these resources, in the case of Pods, may result in consumption of excess resources driving up cost and reducing workload consolidation abilities. The solution to these problems is to identify the derelict resources and remove (i.e., delete) them.
+
+There are two main use cases associated with this.
+
+1. Explicit, per-resource clean-up declaration which may be set by a user either upon creation or retroactively.
```

Comment:
I am concerned about this option - maybe I can't find it but how is implementation planned here?

This sounds (to me) like kyverno would need to detect the addition of a label or annotation on _any_ possible resource type. If this is indeed the intention, then efficient implementation in large clusters could be very hard.
Adding a watch to all possible resources wouldn't be a good idea, but getting the annotation in the minimal timeframe described here (1 minute) also sounds infeasible otherwise.

Maybe I am misunderstand this part of the KDP though.

Reply:
>maybe I can't find it but how is implementation planned here?

Implementation (non-functional requirement) is intentionally open ended here. There are several options. I'm only trying to capture what the outward behavior should be (functional requirement).

>This sounds (to me) like kyverno would need to detect the addition of a label or annotation on any possible resource type. If this is indeed the intention, then efficient implementation in large clusters could be very hard.
Adding a watch to all possible resources wouldn't be a good idea, but getting the annotation in the minimal timeframe described here (1 minute) also sounds infeasible otherwise.

The idea that we (with @realshuting) had here was to use two labels. For example, `kyverno.io/ttl` and `kyverno.io/expires`. With these labels, we can have basically a "static" rule in the ValidatingWebhookConfiguration that uses the `objectSelector` on these labels causing the API server to only forward resources with one of these labels. Once Kyverno can observe this as an AdmissionReview, it can take some action (which is where the implementation details come into play).

Reply:
I hadn't considered the `objectSelector` usage - my only concern left is that this implies that Kyverno would register additional webhooks in the kubernetes API and therefor possibly get events multiple times from different registered webhooks. Not sure how big that problem really is though - maybe something to be discovered during testing.

Reply:
It technically wouldn't even need to be a separate ValidatingWebhookConfiguration; it could just be a separate entry in the existing one. I think we'll have to work through these details when it gets implemented.